torch/distributed/algorithms/ddp_comm_hooks/ddp_zero_hook.py

```diff
@@ -0,0 +1,110 @@
+            modify.
+        ddp (DistributedDataParallel): the :class:`DistributedDataParallel`
+            instance to use.
+        zero (ZeroRedundancyOptimizer): the :class:`ZeroRedundancyOptimizer`
+            instance to use.
+
+    Returns:
+        The modified hook.
+    """
+    assert zero._is_functional_optim, "ZeroRedundancyOptimizer must be " \
+        "constructed with a functional optimizer class to use this function"
+
+    def hook_then_zero_fn(
+        state,
+        bucket: dist.GradBucket,
+    ) -> torch.futures.Future:
+        r"""
+        Returns a :class:`Future` that gives a gradient bucket tensor and
+        performs a partial :class:`ZeroRedundancyOptimizer` :meth:`step` using
+        the gradients in that bucket.
+
+        Arguments:
+            state: any state for the hook.
+            bucket (dist.GradBucket): the :class:`DistributedDataParallel`
+                gradient bucket.
+        """
+        fut = hook(state, bucket)
+
+        def zero_step(_) -> List[torch.Tensor]:
+            r"""
+            Performs a partial :class:`ZeroRedundancyOptimizer` :meth:`step`
+            using the gradients in the given :class:`DistributedDataParallel`
+            gradient bucket.
+            """
+            # Proceed as normal until the DDP buckets have been rebuilt
+            if not ddp._has_rebuilt_buckets:
+                return [bucket.get_tensor()]
+
+            bucket_index = bucket.get_index()
+            rank = zero.global_rank
+            overlap_info = zero._overlap_info
+            if overlap_info.status == _OverlapStatus.UNINITIALIZED:
+                overlap_info.status = _OverlapStatus.DDP_HAS_REBUILT_BUCKETS
+
+            bucket_params = bucket.get_model_params_for_bucket()
+            assert len(bucket_params) > 0, "Empty bucket"
+            rank_to_update = zero._ddp_bucket_index_to_rank(bucket_index)
+
+            # Once DDP buckets have been rebuilt but ZeRO has not been
+            # properly initialized yet, collect the information needed
+            if overlap_info.status == _OverlapStatus.DDP_HAS_REBUILT_BUCKETS:
+                params_per_rank = overlap_info.params_per_rank
+                params_per_bucket = overlap_info.params_per_bucket
+                if rank_to_update == rank:
+                    overlap_info.offsets[bucket_index] = len(params_per_rank[rank_to_update])
+                params_per_rank[rank_to_update].extend(bucket_params)
+                params_per_bucket.append(bucket_params)
+
+                return [bucket.get_tensor()]
+
+            if rank_to_update == rank:
+                assert len(zero.optim.param_groups) == 1, \
+                    "Overlapping DDP with ZeRO only supports a single " \
+                    "parameter group"
+                # Construct the `gradients` input for the local optimizer step,
+                # which expects `None` in a list position to indicate that the
```

Comment:
That might be a little bit brittle or easy for future developer to forget, can you do something like have a global:

_ZERO_NO_UPDATE_PARAM = None 

And use that constant instead, might make the intent more clear.